syntax = "proto3";

package dashboard.v1;

enum FirmwareUpdateStage {
  FIRMWARE_UPDATE_STAGE_IDLE = 0;
  FIRMWARE_UPDATE_STAGE_DOWNLOADING = 1;
  FIRMWARE_UPDATE_STAGE_VERIFYING = 2;
  FIRMWARE_UPDATE_STAGE_REQUESTING_DFU = 3;
  FIRMWARE_UPDATE_STAGE_FLASHING = 4;
  FIRMWARE_UPDATE_STAGE_VERIFYING_VERSION = 5;
  FIRMWARE_UPDATE_STAGE_COMPLETED = 6;
  FIRMWARE_UPDATE_STAGE_FAILED = 7;
  FIRMWARE_UPDATE_STAGE_ROLLING_BACK = 8;
  FIRMWARE_UPDATE_STAGE_ROLLED_BACK = 9;
  FIRMWARE_UPDATE_STAGE_CANCELED = 10;
}

message FirmwareUpdateStatus {
  FirmwareUpdateStage stage = 1;
  float progress = 2;
  string message = 3;
  bool active = 4;
  string current_version = 5;
  string target_version = 6;
  string last_error = 7;
  bool rollback_available = 8;
  uint64 started_at_ns = 9;
  uint64 updated_at_ns = 10;
  uint32 mcu_fw_version_raw = 11;
  uint32 mcu_fw_build = 12;
}

message Status {
  enum State {
    STATE_INIT = 0;
    STATE_IDLE = 1;
    STATE_ACTIVE = 2;
    STATE_FAULT = 3;
  }
  enum CalibrationState {
    CALIBRATION_UNKNOWN = 0;
    CALIBRATION_IDLE = 1;
    CALIBRATION_RUNNING = 2;
    CALIBRATION_PASSED = 3;
    CALIBRATION_FAILED = 4;
  }
  State state = 1;
  bool estop_active = 2;
  bool session_active = 3;
  string active_profile = 4;
  string session_id = 5;
  string last_error = 6;
  uint64 updated_at_ns = 7;
  CalibrationState calibration_state = 8;
  float calibration_progress = 9;
  string calibration_message = 10;
  uint32 calibration_attempts = 11;
  uint64 last_calibration_at_ns = 12;
  FirmwareUpdateStatus firmware_update = 13;
  string mcu_firmware_version = 14;
  uint32 mcu_firmware_build = 15;
  uint32 mcu_update_state = 16;
  uint32 mcu_update_result = 17;
}

message GetStatusRequest {}
message GetStatusResponse { Status status = 1; }

message CalibrateRequest { string profile_id = 1; }
message CalibrateResponse { bool ok = 1; string message = 2; }

message CancelCalibrationRequest {}
message CancelCalibrationResponse { bool ok = 1; string message = 2; }

message SetProfileRequest { string profile_id = 1; }
message SetProfileResponse { bool ok = 1; string active_profile = 2; }

message EStopRequest { bool engaged = 1; string reason = 2; }
message EStopResponse { bool ok = 1; }

message StartSessionRequest {
  string session_id = 1;
  string track = 2;
  string car = 3;
  uint64 start_time_ns = 4;
}
message StartSessionResponse { bool ok = 1; }

message EndSessionRequest { string session_id = 1; }
message EndSessionResponse { bool ok = 1; }

message SessionMetadata {
  string session_id = 1;
  string track = 2;
  string car = 3;
  uint64 start_time_ns = 4;
  uint64 end_time_ns = 5;
  uint64 duration_ms = 6;
}

message ListSessionsRequest {}
message ListSessionsResponse { repeated SessionMetadata sessions = 1; }

message TelemetrySample {
  uint64 timestamp_ns = 1;
  float pitch_rad = 2;
  float roll_rad = 3;
  float left_target_m = 4;
  float right_target_m = 5;
  float latency_ms = 6;
  float speed_kmh = 7;
  int32 gear = 8;
  float engine_rpm = 9;
  float track_progress = 10;
}

message TelemetryStreamRequest { string session_id = 1; }

message InputEventStreamRequest {}

message InputEvent {
  enum Type {
    INPUT_EVENT_TYPE_UNKNOWN = 0;
    INPUT_EVENT_TYPE_PTT_DOWN = 1;
    INPUT_EVENT_TYPE_PTT_UP = 2;
  }
  enum Source {
    INPUT_EVENT_SOURCE_UNKNOWN = 0;
    INPUT_EVENT_SOURCE_STEERING_WHEEL = 1;
  }

  uint64 sequence = 1;
  Type type = 2;
  Source source = 3;
  uint64 received_at_ns = 4;
  uint32 mcu_uptime_ms = 5;
  bool pressed = 6;
}

message GetSessionTelemetryRequest {
  string session_id = 1;
  uint32 max_samples = 2;
}
message GetSessionTelemetryResponse { repeated TelemetrySample samples = 1; }

message StartFirmwareUpdateRequest {
  string artifact_uri = 1;
  string sha256 = 2;
  string target_version = 3;
  bool allow_rollback = 4;
  string rollback_artifact_uri = 5;
}

message StartFirmwareUpdateResponse {
  bool ok = 1;
  string message = 2;
  FirmwareUpdateStatus status = 3;
}

message CancelFirmwareUpdateRequest {}

message CancelFirmwareUpdateResponse {
  bool ok = 1;
  string message = 2;
  FirmwareUpdateStatus status = 3;
}

message CheckFirmwareVersionRequest { string latest_version = 1; }

message CheckFirmwareVersionResponse {
  bool ok = 1;
  bool update_available = 2;
  string current_version = 3;
  string latest_version = 4;
  string message = 5;
  bool mcu_connected = 6;
}

service DashboardService {
  rpc GetStatus(GetStatusRequest) returns (GetStatusResponse);
  rpc Calibrate(CalibrateRequest) returns (CalibrateResponse);
  rpc CancelCalibration(CancelCalibrationRequest) returns (CancelCalibrationResponse);
  rpc SetProfile(SetProfileRequest) returns (SetProfileResponse);
  rpc EStop(EStopRequest) returns (EStopResponse);
  rpc StartSession(StartSessionRequest) returns (StartSessionResponse);
  rpc EndSession(EndSessionRequest) returns (EndSessionResponse);
  rpc ListSessions(ListSessionsRequest) returns (ListSessionsResponse);
  rpc StreamTelemetry(TelemetryStreamRequest) returns (stream TelemetrySample);
  rpc StreamInputEvents(InputEventStreamRequest) returns (stream InputEvent);
  rpc GetSessionTelemetry(GetSessionTelemetryRequest) returns (GetSessionTelemetryResponse);
  rpc StartFirmwareUpdate(StartFirmwareUpdateRequest) returns (StartFirmwareUpdateResponse);
  rpc CancelFirmwareUpdate(CancelFirmwareUpdateRequest) returns (CancelFirmwareUpdateResponse);
  rpc CheckFirmwareVersion(CheckFirmwareVersionRequest) returns (CheckFirmwareVersionResponse);
}
